### PR TITLE
Implement ViewHeader component, make view structure consistent (#801)

### PR DESCRIPTION
**Pull request by ssciolla**
This PR works to make the titles consistent across all the views and to announce the view title programmatically by sending focus to that element. More details are provided below in the task list. The PR aims to resolve issue #801 and work toward resolving issue #807.

- [x] Create `ViewHeader` component that sends focus to the element on initial render.
   - [x] Use `onBlur` event handler to trigger removal of `tabIndex` attribute when focus moves away from header.
   - [x] Add custom CSS class for modifying focus offset (Note: don't think modification of `focus` is possible without CSS or some other helper library; we can also consider modifying the focus style more or remove it entirely if we think it's too distracting).
- [x] Implement `ViewHeader` in "ResourcesAccessed" (required minor edits).
- [x] Implement `ViewHeader` in "Grade Distribution".
   - [x] Use `ViewHeader`. 
   - [x] Refactor slightly so the header will always be visible, even when the `AlertBanner` indicates there is no grade data available.
- [x] Implement `ViewHeader` in "Assignment Planning" (version 1).
   - [x] Refactor so that all page elements are within a single `Paper` component` (they were split between two before).
   - [x] Use `ViewHeader`.
   - [x] Add a new `section` class with top and bottom padding.
   - [x] Refactor to add `div`s that can use the `section` class to ensure proper spacing between sections ("Progress toward Final Grade" and "Assignments Due by Date").
- [x] Implement `ViewHeader` in "Assignment Planning" (version 2).
   - [x] Use `ViewHeader`.
   - [x] Add a new `section` class with top and bottom padding.
   - [x] Refactor to add `div`s that can use the `section` class to ensure proper spacing between sections ("Grade Progress" and "Assignments by Due Date").
- [x] Add `variantMapping` to the theme in `globals.js` so that `h5` and `h6` appear as `h1` and `h2` respectively in HTML.

### CODE REVIEW

**Comment by ssciolla** [comment]
I am hoping to chat with an ITS accessibility expert about the focusing part of these changes, but I think the other changes to view structure and spacing will increase consistency. Unfortunately some of these changes meant I had to large chunks of the JSX for these views (mostly adding `div` tags and indenting), so some significant testing may be required.

**Comment by pushyamig** [comment]
@ssciolla , I think you can put your code on the dev instance and ask for a review from Gonzalo(@gsilver)when this PR opens. He can give you some insights into it. He has been removed from our github account but if it is useful I can add him as colloborator

**Comment by ssciolla** [comment]
> @ssciolla , I think you can put your code on the dev instance and ask for a review from Gonzalo(@gsilver)when this PR opens. He can give you some insights into it. He has been removed from our github account but if it is useful I can add him as colloborator

Thanks, @pushyamig, I already reached out to Gonzalo by email. I will plan to deploy this to the dev instance if I meet with him, or at some point during the review process.

File: assets/src/components/ViewHeader.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+  // Focuses the page header only when the domElement changes.
+  // This should only be when it changes from null to something on initial render.
+  useEffect(() => {
+    if (domElement !== null) {
+      domElement.focus()
+    }
+  }, [domElement])
+
+  return (
+    <div tabIndex={0} ref={domElement => setDomElement(domElement)}>
```
Comment (ssciolla) [inline assets/src/components/ViewHeader.js:17]: If we don't want this `div` to be in the Tab order, but there is some benefit when using a screen reader to focus on this element, we could consider changing this to `-1`. See [MDN's tabindex page](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/tabindex).

**Comment by justin0022** [comment]
@ssciolla I'm just taking an initial look, and I know the intent is for the title to be "focused", but does this mean that the title must have the "blue box" around it? 

Screen Shot 2020-06-03 at 10 12 43 AM

**Comment by ssciolla** [comment]
> @ssciolla I'm just taking an initial look, and I know the intent is for the title to be "focused", but does this mean that the title must have the "blue box" around it?

Yeah, that's why I'm unsure about the current solution. If we change the `tabIndex` to `-1`, I think it would be focused but not available in the Tab order. However, without knowing how to use a screen reader myself, I don't know if that solves the issue raised by the accessibility review. That's why I'm reaching out to the ITS expert.

**Comment by jonespm** [comment]
This doesn't even need a tabindex attribute if there's no action that can be taken on it.

It would have a tabindex=-1 if it should be skipped by keyboard navigation but there should be javascript actions available to it. (Which is rare)

**Comment by ssciolla** [comment]
> Unless it's a link or a form control it shouldn't be tabbable and we should just remove the index. It doesn't _look_ like either of those but I haven't run the code.

@jonespm, we need to send focus to this element, I think, and everything I've read suggests that you need a `tabIndex` to focus. The proposed solution to the original issue was "Change the title on view/route change. **Or pass the focus to a heading that announces the view change.**" 

I don't think this needs to be available via the Tab order, so I've changed it to `-1`, which is what Gonzalo suggested in my initial communications with him. I'm still figuring out how to manage the prominent focus on the title, which feels a little distracting and unnecessary.

**Comment by ssciolla** [comment]
FYI, I've deployed my latest code to the dev MyLA instance if you want to take a look. I haven't made any styling changes yet.

File: assets/src/components/ViewHeader.js [hunk 2]
```diff
@@ -1,43 +1,33 @@
 import React, { useEffect, useState } from 'react'
 import Typography from '@material-ui/core/Typography'
-import useTimeout from '../hooks/useTimeout'
```
Comment (pushyamig) [inline assets/src/components/ViewHeader.js]: This is exactly what I was thinking, why do we need a timeout with all that hook fancy code, looks impressive but is that needed really. Thank for keeping it simple

Comment (ssciolla) [reply assets/src/components/ViewHeader.js]: Yeah, Gonzalo said the timeout wasn't really necessary, I think the use of `onBlur` makes more sense.

File: assets/src/components/ViewHeader.js (hunk 1 above)
Comment (ssciolla) [reply assets/src/components/ViewHeader.js]: I've changed it to `-1` but also I'm removing the `tabIndex` when the focus moves away now.

**Comment by ssciolla** [comment]
K, I think this is ready for review. I'm rebuilding an image for MyLA dev so should be available later today or tomorrow.

File: assets/src/containers/ResourcesAccessed.js [hunk 3]
```diff
@@ -236,7 +237,7 @@ function ResourcesAccessed (props) {
       <Grid container spacing={2}>
         <Grid item xs={12}>
           <Paper className={classes.paper}>
-            <Typography variant='h5' gutterBottom className='title'>Resources Accessed</Typography>
```
Comment (pushyamig) [inline assets/src/containers/ResourcesAccessed.js:239]: Since you removed the Typography can you also remove the import as well?

File: assets/src/components/ViewHeader.js [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+  // This should only be when it changes from null to something on initial render.
+  useEffect(() => {
+    if (domElement !== null) {
+      domElement.focus()
+    }
+  }, [domElement])
+
+  const headerProps = {
+    className: 'view-header',
+    variant: 'h5',
```
Comment (pushyamig) [inline assets/src/components/ViewHeader.js:21]: Since the variant mapping to appears as h5->h1 and h6->h2, does it make sense to have this variant as `h5` instead of `h1`?

File: assets/src/globals.js [hunk 5]
```diff
@@ -20,13 +20,27 @@ const user = Object.freeze({
...
 
-const siteTheme = createMuiTheme({ palette })
+// Variant mapping: https://material-ui.com/components/typography/#changing-the-semantic-element
+// This will make uses of h5 appear as h1 and h6 appear as h2, without changing styling.
+const componentSettings = {
+  MuiTypography: {
+    variantMapping: {
+      h5: 'h1',
+      h6: 'h2'
+    }
```
Comment (pushyamig) [inline assets/src/globals.js:36]: Why is this change needed?

**Comment by pushyamig** [comment]
I looked at the color around the title header and it is a little bit different than the Myla color palette. is there a reason for this color?
Screen Shot 2020-06-05 at 8 28 21 AM

File: assets/src/containers/ResourcesAccessed.js (hunk 3 above)
Comment (ssciolla) [reply assets/src/containers/ResourcesAccessed.js:239]: Oh, yeah, I'll check to see if we can remove that, thanks.

File: assets/src/globals.js (hunk 5 above)
Comment (ssciolla) [reply assets/src/globals.js:36]: If this has been implemented correctly, all uses of `h5` should appear as `h1`, and `h6` as `h2` in the page's HTML. According to Gonzalo, [Material UI](https://material-ui.com/components/typography/#changing-the-semantic-element), and [W3C](https://www.w3.org/WAI/tutorials/page-structure/headings/), it is important for accessibility that a page's headings start at `h1` and descend without skipping levels. Previously, we were using `h5` and `h6` as the preferred styles for headings, so I wanted to maintain those styles while making the heading numbers appear correctly in the HTML. Does that make sense?

File: assets/src/components/ViewHeader.js (hunk 4 above)
Comment (ssciolla) [reply assets/src/components/ViewHeader.js:21]: This has been confusing to me, but I think we have to use the heading with the style we want (`h5`) and `variantMapping` will convert the actual tag to the right number.

**Comment by ssciolla** [comment]
> I looked at the color around the title header and it is a little bit different than the Myla color palette. is there a reason for this color?

The only modification I am making right now to the focus style is to add an offset of `5px` to the `outline` (see the changes to `index.css`), so what you're seeing should be your browser's (or maybe Material UI's?) focus treatment. It doesn't seem like the offset is being registered in the image you showed me for some reason.

@pushyamig, what style would you like this to be? Would you like me to make invisible even? I can style the outline in terms of pattern or color in various different ways, or remove it, just let me know what you think would be best.

**Comment by jonespm** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 2
           

Complexity increasing per file
==============================
- assets/src/components/ViewHeader.js  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ITS_Teaching_And_Learning/my-learning-analytics/pullRequest?prid=5646924&bid=18492831)

File: assets/src/globals.js (hunk 5 above)
Comment (pushyamig) [reply assets/src/globals.js:36]: Indeed make sense, this is a kind global change for this inline version
`<Typography variant="h1" component="h5">`
This is reflected correctly in the UI and it doesn't have the same effect of `<h2>` header tag
Screen Shot 2020-06-05 at 10 41 55 AM

Screen Shot 2020-06-05 at 10 42 55 AM

File: assets/src/containers/ResourcesAccessed.js (hunk 3 above)
Comment (ssciolla) [reply assets/src/containers/ResourcesAccessed.js:239]: This should be resolved now.

File: assets/src/index.css [hunk 6]
```diff
@@ -26,6 +26,11 @@ code {
   animation: sk-scaleout 1.0s infinite ease-in-out;
 }
 
+/* Modifies default focus style for header tag in ViewHeader component */
+.view-header:focus {
+  outline-offset: 5px;
```
Comment (pushyamig) [inline assets/src/index.css:31]: I don't see the effect of this styling in the UI. I have changed the outline-offset setting am I missing something?
Screen Shot 2020-06-05 at 11 04 41 AM

Comment (pushyamig) [reply assets/src/index.css:31]: The `outline-offset: 5px;` property doesn't see the effect the chrome, but only in FF. I am not sure why is this case? Need to support all prime browser as per this [list](https://github.com/tl-its-umich-edu/my-learning-analytics/wiki/MyLA-Browser-Support)

**Comment by pushyamig** [comment]
> what style would you like this to be? Would you like me to make the focus invisible even? I can style the outline in terms of pattern or color in various different ways, or remove it. Just let me know what you think would be best.

Invisible is better, if that is not the expected behavior then add color to the header the blend in with current style guide recommendation

File: assets/src/index.css (hunk 6 above)
Comment (pushyamig) [reply assets/src/index.css:31]:
```
outline-offset: 5px;
  outline-style: solid;
```
This combination seem to work in Chrome to add little spacing around the box

Comment (ssciolla) [reply assets/src/index.css:32]: Okay, I think I added a specific style for the focus so it is consistent across browsers. Below is a screenshot of how it looks in Chrome. I opted to use the `secondary` color (`#2C6496`), since that indicates a selected state in the rest of the application.
Screen Shot 2020-06-05 at 1 24 48 PM

Comment (ssciolla) [reply assets/src/index.css:32]: @jennlove-um, I would like your input on how this looks. What I've seen online suggests that [focus should be visible](https://www.w3.org/TR/UNDERSTANDING-WCAG20/navigation-mechanisms-focus-visible.html), but this could be considered a slightly different case since it is not for an element that accepts user interaction. Gonzalo said that, if we are concerned about the appearance, it would be okay to remove the focus style entirely. It will still focus when the page loads, but it just won't be visible. Note that the focus box (seen above) will disappear if the user presses Tab or otherwise moves the focus elsewhere.

**Comment by jonespm** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 3
           

Complexity increasing per file
==============================
- assets/src/components/ViewHeader.js  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ITS_Teaching_And_Learning/my-learning-analytics/pullRequest?prid=5646924&bid=18492831)

**Comment by jonespm** [comment]
On the[ ReactJS accessibility page](https://reactjs.org/docs/accessibility.html) they mentioned the [DocumentTitle](https://github.com/gaearon/react-document-title) and [React Helmet package](https://github.com/nfl/react-helmet) that sound like they do updates directly to the header. Did you try either of these?

**Comment by ssciolla** [comment]
> On the[ ReactJS accessibility page](https://reactjs.org/docs/accessibility.html) they mentioned the [DocumentTitle](https://github.com/gaearon/react-document-title) and [React Helmet package](https://github.com/nfl/react-helmet) that sound like they do updates directly to the header. Did you try either of these?

I looked into `react-helmet` and raised the possibility of changing the document title this way with Gonzalo. He replied "Changing the title is a tall proposition and may not be useful to all screen reading software. I think the best solution is to send focus to the heading, and remove the tabindex when that loses focus."

I don't think using something like this is a bad idea (although I'm not exactly sure how I would change the title, since it's currently "My Learning Analytics" -- maybe adding something after a colon), but from Gonzalo's perspective, it seems the header focus might be more helpful to users.

**Comment by jonespm** [comment]
Before single page apps, the user noticed that they changed pages by the title changes. I feel like updating the title would have benefit both to visual users as well as screen reader users as this is what's displayed in the tab. 

Apps like Gmail are a SPA and the title updates when you folders something from 

`Inbox - jonespm@gmail.com - Gmail`
`Starred - jonespm@gmail.com - Gmail`
or more specifically

`<Folder Name> <email> - Gmail`

It seems like our style should be of the format
`<Page Name> - <username> - My Learning Analytics`

Sorry for being a day late to the comments here. It's also possible the username is not needed.

**Comment by ssciolla** [comment]
> Before single page apps, the user noticed that they changed pages by the title changes. I feel like updating the title would have benefit both to visual users as well as screen reader users as this is what's displayed in the tab.

@jonespm, yeah, I think this is a good idea. However, with the way you are commenting (without reviewing or requesting changes), I am very unclear about how you want me to proceed. Are you suggesting what I've done with focus should be removed and replaced with something like a `react-helmet` implementation? Are you saying this would be good to in addition? Is it something that has to be in this PR, or could it be resolved in a separate issue? If you -- as a developer senior to myself on this project, and someone whose opinion I value -- could give me clear direction, that would be appreciated.

**Comment by jonespm** [comment]
@ssciolla I don't consider myself to be very experienced with React. I also haven't installed a screenreader to test this out. I've done training and read a bit more and done work on accessibility in general and just making observations based on other pages and also Django Admin on how they change the title.

I don't know anything about those two libraries I mentioned or what they do. I'd estimate would take a few hours of effort to try them out myself. Perhaps they could replace some of this effort, perhaps they could supplement it. I was just wondering if you'd considered them as it looked a little similar to what you were writing with this new component. 

I usually just figure if there's a seemingly common problem that someone else already has a good solution, this didn't feel very unique to this project. 

The code here could probably be swapped out later too for something else.

Sorry there's really no clear direction, as that direction would be:
- Try out these libraries and let us know how they work
- Try out all 3 solutions with a screenreader (VoiceOver on OSX works, NVAccess is free and popular), let us know what works best. I know JAWS is popular but don't know if what is most used anymore.

I tested this with VoiceOver. This solution works better than changing the Title as even on Gmail the title updates don't get Voiceover to read them. I don't know  about other screenreaders. I feel *both* solutions might be useful but not sure of the effort.

**Comment by ssciolla** [comment]
Okay, @jonespm, thanks for explaining a bit more. I am feeling the pressure to at least partially resolve this issue/PR because we are trying to release next week. Because there is research involved in deciding on a library and figuring out it works, I feel it would make sense to tackle the title updating separately. I will create an issue and assign it to myself. It would still be great to have your review on what I have here -- it sounds like it will be, based on your testing and Gonzalo's expertise.

**Comment by pushyamig** [comment]
> I also haven't installed a screenreader to test this out.
I don't want to make the conversations anymore longer than what it is but do we want to make testing with screen readers as a standard process for resolving the accessibility issue? It will be hard to reach out to Accessibility experts for each and every issue.

**Comment by ssciolla** [comment]
FWIW, I just tried implementing title changes with `react-helmet`, and it was quite easy: https://github.com/ssciolla/my-learning-analytics/commit/4c4e6bd6168b8651ee26067da4c71702a860d36c

I still think it might be worth investigating this more, and it would likely make sense to adjust the title for other routes or views, like the list of courses or views. I'm also not sure the best place for `Helmet` is in `ViewHeader`; maybe it should be wrapped around the instances of `Route`. Thus I'd still like to try to tackle this in another issue/PR and limit the scope of this one, since I'm already making a good number of changes.

**Comment by pushyamig** [comment]
We should merge this what @ssciolla has done here is valuable and we are all learning with accessibility stuff with this project. It is separate Question how will you mark this issue as done

**Comment by pushyamig** [comment]
> Thus I'd still like to try to tackle this in another issue/PR and limit the scope of this one, since I'm already making a good number of changes.

I agree

**Comment by ssciolla** [comment]
I spoke with @jennlove-um about the focus styling, and we opted to remove the focus outline (i.e. change the CSS class to only use `outline: none`. I inspected the HTML and used VoiceOver, and the title still properly receives focus.

**Comment by jonespm** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 2
           

Complexity increasing per file
==============================
- assets/src/components/ViewHeader.js  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ITS_Teaching_And_Learning/my-learning-analytics/pullRequest?prid=5646924&bid=18492831)

**Comment by ssciolla** [comment]
I'm merging now, thanks everyone for the feedback.
